### PR TITLE
New draw transform

### PR DESCRIPTION
**Pull request by albi3ro**
This PR updates the `qml.draw` transform to use the new `qml.drawer.tape_text` function. The old transform can be accessed by `qml.transforms.draw_old`.

## Differences

The new function uses a different algorithm to place the operations into layers, see `qml.drawer.drawable_layers` versus `CircuitGraph.greedy_layers`. This can result in a different placement of operations.

We now assume Unicode support, given how universal Unicode support is.  Therefore the new draw does not have a `charset` keyword.

Parameters are now omitted by default.  The users can configure the display resolution of parameters via the `decimals` keyword.

Spacing is slightly different.

`wire_order` can now contain wires not on the device's wires.  Any wire acted on by operations but not present in the `wire_order` will be placed below everything in the `wire_order`.

The new drawer restricts to fix-width characters, avoiding the use of characters like ⊗ and ⟩, as those can mess up the alignment.

## Example

```python
dev = qml.device('default.qubit', wires=["a", -1, "q2"])

@qml.qnode(dev)
def circuit():
    qml.Hadamard(wires=-1)
    qml.CNOT(wires=["a", "q2"])
    qml.RX(0.2, wires="a")
    return qml.expval(qml.PauliX(wires="q2"))
```

```
>>> print(qml.draw(circuit)())
 a: ────╭C──RX─┤     
-1: ──H─│──────┤     
q2: ────╰X─────┤  <X>
>>> print(qml.transforms.draw_old(circuit)())
  a: ─────╭C──RX(0.2)──┤     
 -1: ──H──│────────────┤     
 q2: ─────╰X───────────┤ ⟨X⟩ 
```

There are failing tests due to the old qnode, which will likely be removed in the next release anyway.

### CODE REVIEW

File: pennylane/transforms/draw.py [hunk 1]
```diff
@@ -152,6 +152,148 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+def draw(
+    qnode,
+    wire_order=None,
+    show_all_wires=False,
+    decimals=None,
```
Comment (dime10) [inline pennylane/transforms/draw.py:159]: Could two be the default here? I had the impression people expect to see the rotation angles, and it helps differentiate the different rotations appearing in the circuit.

File: pennylane/transforms/draw.py [hunk 2]
```diff
@@ -152,6 +152,148 @@ def wrapper(*args, **kwargs):
...
+
+    >>> print(qml.draw(circuit)(a=2.3, w=[1.2, 3.2, 0.7]))
+    0: ──H─╭C───────╭C──┤ ╭<Z@Z>
+    1: ────╰RX──Rot─╰RX─┤ ╰<Z@Z>
+
+    .. UsageDetails::
+
+
+    By default, parameters are omitted. By specifying the ``decimals`` keyword, parameters
+    are displayed to the specified precision. Matrix-valued parameters are never displayed.
```
Comment (dime10) [inline pennylane/transforms/draw.py:207]: Here also it might be helpful not necessarily printing the matrix but at least label it M1, M2, etc which helps distinguish the different gates and identify which ones use the same matrix parameter.

File: pennylane/transforms/draw.py [hunk 3]
```diff
@@ -152,6 +152,148 @@ def wrapper(*args, **kwargs):
...
+                tape_text(
+                    t,
+                    wire_order=_wire_order,
+                    show_all_wires=show_all_wires,
+                    decimals=decimals,
+                    max_length=max_length,
+                    cache=cache,
+                )
+                for t in tapes[0]
+            ]
```
Comment (dime10) [inline pennylane/transforms/draw.py:283]: Out of curiosity what scenario is this intended for?

File: tests/transforms/test_draw.py [hunk 4]
```diff
@@ -1,593 +1,242 @@
...
+
+    def test_wire_order(self):
+        """Test wire_order keyword changes order of the wires."""
+
+        split_str = draw(circuit, wire_order=[1.234, "a", 0, "b"])(1.2, 2.3, 3.4).split("\n")
+        assert split_str[0][0:6] == "1.234:"
+        assert split_str[1][0:6] == "    a:"
+        assert split_str[2][0:6] == "    0:"
+
+    def test_show_all_wires(self):
```
Comment (dime10) [inline tests/transforms/test_draw.py:53]: Could this test include the case where extra wires are present from the wire_order argument?

File: tests/transforms/test_draw.py [hunk 5]
```diff
@@ -1,593 +1,242 @@
...
+        expected = "0: ──Rot(1.2,2.3,3.5)─┤  <Z>"
+        assert draw(circuit, decimals=1)([1.234, 2.345, 3.456]) == expected
+
+    def test_decimals_0(self):
+        """Test decimals=0 rounds to integers."""
+
+        expected = "    0: ──RX(1)─┤  <Z>\n    a: ──RY(2)─┤     \n1.234: ──RZ(3)─┤     "
+        assert draw(circuit, decimals=0)(1.234, 2.3456, 3.456) == expected
+
+    def test_torch_parameters(self):
```
Comment (dime10) [inline tests/transforms/test_draw.py:91]: I don't know if terribly relevant but numpy arrays are missing here.

File: tests/transforms/test_draw.py [hunk 6]
```diff
@@ -1,593 +1,242 @@
...
+        """Test max length default to 100."""
+
+        @qml.qnode(qml.device("default.qubit", wires=1))
+        def long_circuit():
+            for _ in range(100):
+                qml.PauliX(0)
+            return qml.expval(qml.PauliZ(0))
+
+        out = draw(long_circuit)()
+        assert 95 <= max(len(s) for s in out.split("\n")) <= 100
```
Comment (dime10) [inline tests/transforms/test_draw.py:129]: Why the lower bound of 95?

File: tests/transforms/test_draw.py [hunk 7]
```diff
@@ -1,593 +1,242 @@
...
+        qml.Hadamard(wires=0)
+        qml.RX(x, wires=0)
+        return qml.expval(qml.PauliZ(0))
+
+    expected = "0: ──H──RX(0.8)─┤  <Z>\n\n0: ──H──RX(0.4)─┤  <Z>"
+    assert draw(circuit, decimals=1)(np.array(0.6, requires_grad=True)) == expected
+
+
+def test_nested_tapes():
+    """Test nested tapes inside the qnode."""
```
Comment (dime10) [inline tests/transforms/test_draw.py:200]: 💯

**Review by dime10 — CHANGES_REQUESTED** [review]
Looks great, thanks @albi3ro ! 🎉

File: pennylane/transforms/draw.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/transforms/draw.py:283]: batch transforms.

File: pennylane/transforms/draw.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/transforms/draw.py:207]: I have a plan to add support for this in the future. If necessary, I can wait on this PR till  I figure that out.

Comment (dime10) [reply pennylane/transforms/draw.py:207]: I'm ok with separate PRs, that way things keep moving forward :)

File: pennylane/transforms/draw.py (hunk 3 above)
Comment (dime10) [reply pennylane/transforms/draw.py:283]: Oh that's awesome, I didn't realise either that the previous drawer could do that as well. Do you think it would be a worthwhile example for the docstring?

**Comment by albi3ro** [comment]
[sc-12940]

File: tests/transforms/test_draw.py [hunk 8]
```diff
@@ -1,593 +1,242 @@
...
+        @qml.qnode(qml.device("default.qubit", wires=(0, 1)))
+        def circuit():
+            return qml.expval(qml.PauliZ(0))
+
+        split_str = draw(circuit, show_all_wires=True)().split("\n")
+        assert split_str[0][0:2] == "0:"
+        assert split_str[1][0:2] == "1:"
+
+
+class TestDecimals:
```
Comment (Jaybsoni) [inline tests/transforms/test_draw.py:65]: might be excessive, but having a test where decimals provides more precision then needed for the parameters?
 `draw(circuit, decimals=2)(1,2,3)` 
such that the result is 
`--RX(1.00)--RY(2.00)--RZ(3.00)--|`

**Review by Jaybsoni — COMMENTED** [review]
Looks good! Not much to update!

File: pennylane/drawer/tape_text.py [hunk 9]
```diff
@@ -252,7 +252,7 @@ def tape_text(
                 finished_lines += totals
                 finished_lines[-1] += "\n"
                 totals = [filler] * n_wires
-                line_length = 1 + max_label_len
+                line_length = 2 + max_label_len
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/drawer/tape_text.py:255]: In testing, I determined that in some cases were actually longer than the specified `max_length`.  This was the source of that bug.

**Comment by dime10** [comment]
@albi3ro could mark or reply to the comments you've addressed?

File: tests/transforms/test_draw.py (hunk 4 above)
Comment (albi3ro) [reply tests/transforms/test_draw.py:53]: added below.

File: tests/transforms/test_draw.py (hunk 8 above)
Comment (albi3ro) [reply tests/transforms/test_draw.py:76]: added. `test_decimals_higher_value`.

File: tests/transforms/test_draw.py (hunk 5 above)
Comment (albi3ro) [reply tests/transforms/test_draw.py:119]: `test_qml_numpy_parameters` above. looks like I missed a docstring... oops

File: tests/transforms/test_draw.py (hunk 6 above)
Comment (dime10) [reply tests/transforms/test_draw.py:163]: Oh I see 👍

**Review by dime10 — APPROVED** [review]
Fantastic! Looks good to me once the failing tests are fixed 💯

File: doc/releases/changelog-dev.md [hunk 10]
```diff
@@ -4,6 +4,26 @@
...
+      qml.CRX(a, wires=[0, 1])
+      qml.Rot(*w, wires=[1])
+      qml.CRX(-a, wires=[0, 1])
+      return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+  ```
+  ```
+    >>> print(qml.draw(circuit, decimals=2)(a=2.3, w=[1.2, 3.2, 0.7]))
+    0: ──H─╭C─────────────────────────────╭C─────────┤ ╭<Z@Z>
+    1: ────╰RX(2.30)──Rot(1.20,3.20,0.70)─╰RX(-2.30)─┤ ╰<Z@Z>
+  ```
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:25]:
```suggestion
  >>> print(qml.draw(circuit, decimals=2)(a=2.3, w=[1.2, 3.2, 0.7]))
  0: ──H─╭C─────────────────────────────╭C─────────┤ ╭<Z@Z>
  1: ────╰RX(2.30)──Rot(1.20,3.20,0.70)─╰RX(-2.30)─┤ ╰<Z@Z>
```

**Review by Jaybsoni — APPROVED** [review]
